use-cases/Separate-interface-from-implementation.md

```diff
@@ -0,0 +1,29 @@
+# Separate interface from implementation
+
+## General information
+
+The interface and implementation of package definitions should be separated to ensure better maintainability, flexibility, and compatibility between different tools and approaches. A well-defined interface would allow package maintainers to focus on the implementation details while providing stability and consistency for end-users. It would also enable out-of-tree packaging libraries to interoperate with Nixpkgs more effectively, reducing the need for duplicated efforts and custom workarounds.
+
```

Comment:
Something to be addressed: while having clear interfaces is obviously good for maintainability, it is also a pretty powerful feature of Nixpkgs that you *can* override things in a hacky way, e.g. through `.overrideDerivation`. I'm not sure we want to disallow such mechanisms entirely, since they're often the most convenient way to do local overrides. The alternative would be having to clone the package expression.

So we may want to still provide hacky override mechanisms like `.overrideDerivation`, as long as it's clear to users that such mechanisms don't come with any interface stability guarantees.

Reply:
Re-phrased this slightly and added it to the general information section via 7b6081fce85129c8d82afb2e92af5ecc615b6a3c